### PR TITLE
[WIP] Allow F expression to function argument

### PR DESCRIPTION
**Pull request by lntuition**
## Description
This patch is related #257. Allowing `F` expression to `Function` argument. 

* [x] Allow  `F` expression to `Function` argument
* [x] Fix type casting (with broadcasting also)
* [x] Add testcase (including negative test)
* [x] Add change to CHANGELOG.rst

Current :
Waiting review...

### CODE REVIEW

File: tortoise/queryset.py [hunk 1]
```diff
@@ -667,9 +667,7 @@ def _make_query(self) -> None:
                 if not isinstance(value, (F, ArithmeticExpression)):
```
Comment (grigi) [inline tortoise/queryset.py]: :thinking:  Should we change this to check for instance type `pypika.Term`? That is the common base class of both those?

File: tortoise/functions.py [hunk 2]
```diff
@@ -34,14 +35,15 @@ class Function:
...
+    __slots__ = ("field", "field_is_str", "field_object", "default_values")
 
     database_func = BaseFunction
     # Enable populate_field_object where we want to try and preserve the field type.
     populate_field_object = False
 
-    def __init__(self, field: str, *default_values: Any) -> None:
+    def __init__(self, field: Union[str, F, ArithmeticExpression], *default_values: Any) -> None:
         self.field = field
+        self.field_is_str: bool = isinstance(field, str)
```
Comment (grigi) [inline tortoise/functions.py]: Is this needed? We are not mutating `self.field` that I can see? Are you planning to in the future?

File: tortoise/expressions.py [hunk 3]
```diff
@@ -1,32 +1,67 @@
...
+                    field_object = left_field_object
+
+            (
+                arithmetic_expression_or_field.right,
+                right_field_object,
+            ) = cls.resolver_arithmetic_expression(model, right)
+            if right_field_object:
+                if field_object and type(field_object) != type(right_field_object):
+                    raise FieldError(
+                        "Cannot use arithmetic expression between different field type"
```
Comment (grigi) [inline tortoise/expressions.py]: :+1:

File: tests/fields/test_decimal.py [hunk 4]
```diff
@@ -95,6 +95,53 @@ def test_decimal_places_bad(self):
...
+        with self.assertRaisesRegex(
+            FieldError,
+            "Field not_exist is not default field in model, so can not be used with F expression",
+        ):
+            await testmodels.DecimalFields.all().annotate(sum_decimal=Sum(F("not_exist"))).values(
+                "sum_decimal"
+            )
+
+        with self.assertRaisesRegex(
+            FieldError, "Cannot use arithmetic expression between different field type"
```
Comment (grigi) [inline tests/fields/test_decimal.py]: Could you split these tests? There is no reason to have them both in the same test?

File: tortoise/expressions.py [hunk 5]
```diff
@@ -1,32 +1,67 @@
...
+                    if func:
+                        arithmetic_expression_or_field = func(
+                            field_object, arithmetic_expression_or_field
+                        )
             except KeyError:
-                raise FieldError(f"Field {name} is virtual and can not be updated")
-            arithmetic_expression_or_field.name = name
+                raise FieldError(
+                    f"Field {name} is not default field in model,"
+                    f" so can not be used with F expression"
```
Comment (grigi) [inline tortoise/expressions.py]: The message here is confusing to me? Do you say this only works with fields that have a default set?

File: tests/fields/test_decimal.py [hunk 6]
```diff
@@ -108,6 +155,37 @@ def test_decimal_places_bad(self):
...
+            .annotate(avg_decimal=Avg(F("decimal") + 1))
+            .values("avg_decimal")
+        )
+        self.assertEqual(
+            values[0], {"avg_decimal": Decimal("13.42")},
+        )
+
+        values = (
+            await testmodels.DecimalFields.all()
+            .annotate(avg_decimal=Avg(F("decimal") + F("decimal")))
```
Comment (grigi) [inline tests/fields/test_decimal.py]: Could you do maths with one of the other fields? `decimal_nodec` or `decimal_null` instead. So we see that it works with both.

Comment (grigi) [reply tests/fields/test_decimal.py]: Also, what is our expected behaviour when doing arithmetic on nulls? We probably want to ensure consistent behaviour at least?

**Comment by grigi** [comment]
@lntuition Looks good! I only have minor comments/questions.

File: tests/fields/test_decimal.py (hunk 4 above)
Comment (lntuition) [reply tests/fields/test_decimal.py]: Sure. I will split those.

File: tortoise/queryset.py (hunk 1 above)
Comment (lntuition) [reply tortoise/queryset.py]: Looks okay. I changed it !

File: tortoise/functions.py (hunk 2 above)
Comment (lntuition) [reply tortoise/functions.py]: Yeah. We are not mutating `self.field` current stage. hmm.. also can removed.. I changed it !

File: tortoise/expressions.py (hunk 5 above)
Comment (lntuition) [reply tortoise/expressions.py]: Yeah. Maybe this message can confuse user. Can you recommend more clearer message?

File: tests/fields/test_decimal.py (hunk 6 above)
Comment (lntuition) [reply tests/fields/test_decimal.py]: We can do with one of the other fields. i will add testcase to test that scenario.

Comment (lntuition) [reply tests/fields/test_decimal.py]: And doing arithmetic on nulls should generate null. But i'm not sure that all database(sqlite, postgres. etc) do equals.

Comment (grigi) [reply tests/fields/test_decimal.py]: About nulls, I feel the most correct behaviour would to have the query fail? And then you'd have to explicitly set a defualt-if-null with `coalese()`?

At least it would be good to run the test with all db's to see what they do? And hope they do the same thing... At minumim just to document it?

File: tortoise/expressions.py (hunk 5 above)
Comment (grigi) [reply tortoise/expressions.py]: We'll only get a `KeyError` if the field is not a DB column. So maybe we should explicitly mention that?
`There is no DB-backed field called {name}`?
or
`There is no non-virtual field {name} on Model {model.__name__}`?

File: tests/fields/test_decimal.py (hunk 6 above)
Comment (lntuition) [reply tests/fields/test_decimal.py]: Could you see this [link](https://www.sqlite.org/nulls.html)? This links saids add anything to null is correct behavior. It should be okay just addding testcase with add something and null. 

Also i add test with add something with null. Does CI test it with all supported database? (e.g sqlite, postgres, etc)

Comment (grigi) [reply tests/fields/test_decimal.py]: CI tests with all supported DB's yes. That's a handy link. It looks like the supported DB's should behave the same.

File: tortoise/expressions.py (hunk 5 above)
Comment (lntuition) [reply tortoise/expressions.py]: I think below looks better ! I will change the message :+1:

File: CHANGELOG.rst [hunk 7]
```diff
@@ -74,6 +74,7 @@ future
 
 - Fix default type of ``JSONField``
 - Install on Windows does not require a C compiler any more.
+- Allow F expression to ``Function``. (#301)
```
Comment (grigi) [inline CHANGELOG.rst]: Could you please expand this a bit. Whilst technically correct, the average user will not see the implication.
Rather say ``Allow usage of F expressions in annotations`` ?

**Comment by grigi** [comment]
Looks good. Please just improve the changelog, and I'll merge. :+1:

File: CHANGELOG.rst (hunk 7 above)
Comment (lntuition) [reply CHANGELOG.rst]: Sure, I change changelog and also resolve conflicts :smile:

**Comment by grigi** [comment]
:+1:
